kanto_kuska/edge_client.py

```diff
@@ -0,0 +1,150 @@
+import time
+import json
+import paho.mqtt.client as mqtt
+
+# Mock Ditto and Kuksa client interactions
+class MockDittoClient:
+    def __init__(self):
+        self.device_info = None
+        self.features = {}
+
+    def send(self, cmd_envelope):
+        feature_id, property_path, value = cmd_envelope.split(":")
+        self.features[f"{feature_id}/{property_path}"] = value
+        print(f"Mock Ditto: Updated feature '{feature_id}' property '{property_path}' with value '{value}'")
+
+class MockKuksaClient:
+    def __init__(self):
+        self.values = {
+            'Altitude': 100,
+            'Latitude': 17.3850,
+            'Longitude': 78.4667,
+            'Speed': 50
+        }
+        self.subscriptions = {}
+        self.update_interval = 1  # Seconds between simulated data updates
+
+    def subscribe(self, topic, callback):
+        print(f"Mock Kuksa subscribed to topic: {topic}")
+        self.subscriptions[topic] = callback
+        callback(self.values[topic])
+
+    def simulate_updates(self):
+        while True:
+            # Simulate changes in some values
+            self.values['Speed'] += 2
+            # Update subscribed callbacks
+            for topic, callback in self.subscriptions.items():
+                callback(self.values[topic])
+            time.sleep(self.update_interval)
+
+class EdgeClient:
+    def __init__(self, mqtt_host="localhost", mqtt_port=1883, mqtt_username=None, mqtt_password=None):
+        self.ditto_client = MockDittoClient()
+        self.kuksa_client = MockKuksaClient()
+        self.device_info = None
+
+        # MQTT settings
+        self.mqtt_host = mqtt_host
+        self.mqtt_port = mqtt_port
+        self.mqtt_username = mqtt_username
+        self.mqtt_password = mqtt_password
+        self.mqtt_client = mqtt.Client()

```

Comment:
When I tried to run the script it failed with:
```
Exception ignored in: <function Client.__del__ at 0x7f4007213e20>
Traceback (most recent call last):
  File "/home/did1sf/.local/lib/python3.10/site-packages/paho/mqtt/client.py", line 874, in __del__
    self._reset_sockets()
  File "/home/did1sf/.local/lib/python3.10/site-packages/paho/mqtt/client.py", line 1133, in _reset_sockets
    self._sock_close()
  File "/home/did1sf/.local/lib/python3.10/site-packages/paho/mqtt/client.py", line 1119, in _sock_close
    if not self._sock:
AttributeError: 'Client' object has no attribute '_sock'
Traceback (most recent call last):
  File "/home/did1sf/develop/github/bosch-io/kanto-example-applications/kanto-kuksa/edge_client.py", line 149, in <module>
    edge_client = EdgeClient(mqtt_host, mqtt_port, mqtt_username, mqtt_password)
  File "/home/did1sf/develop/github/bosch-io/kanto-example-applications/kanto-kuksa/edge_client.py", line 52, in __init__
    self.mqtt_client = mqtt.Client()
TypeError: Client.__init__() missing 1 required positional argument: 'callback_api_version'
```

This is due to paho-mqtt version I have installed - 2.0.0. Check [here](https://github.com/eclipse/paho.mqtt.python/issues/814) for more information. 
This fixed it for me: `self.mqtt_client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION1)`